docs/adr/0001-add-generic-fetcher.md

```diff
@@ -0,0 +1,109 @@
+
+## Decision
+
+This change introduces a generic fetcher, an additional cachi2 package manager. This package manager utilizes a custom
+lockfile that is located in the input repository. Based on that lockfile, it will download files, save them into a requested
+location, and verify checksums. Below is a more detailed overview of the implementation.
+
+### Lockfile format
+
+Cachi2 expects the lockfile to be named `generic_lockfile.yaml`.
+In order to account for possible future breaking changes, the lockfile will contain a `metadata` section with a `version`
+field that will indicate the version of the lockfile format. It will also contain a list of artifacts (files) to download,
+each of the artifacts to have a URL, list of checksums, and optionally target location specified.
+
+```yaml
+metadata:
+  # uses X.Y semantic versioning
+  version: "1.0"
+artifacts:
+  - download_url: https://huggingface.co/instructlab/granite-7b-lab/resolve/main/model-00001-of-00003.safetensors?download=true
+    target: granite-model-1.safetensors
+    checksums:
+        sha256: d16bf783cb6670f7f692ad7d6885ab957c63cfc1b9649bc4a3ba1cfbdfd5230c
+```
+
+#### Lockfile properties
+
+Below is an explanation of individual properties of the lockfile.
+
+##### download_url (required)
+
+Specified as a string containing the download url of the artifact.
+
+##### checksums (required)
+
+Specified as a dictionary of checksum algorithms and their values. At least one cachi2-verifiable checksum must be provided
+to ensure at least some degree of confidence in the identity of the artifact.
+
+#### target (optional)
+
+This key is provided mainly for the users convenience, so the files end up in expected locations. It is optional and if
+not specified, it will be derived from the download_url. Target here means a specific subdirectory inside cachi2's output
+directory for the generic fetcher (`{cachi2-output-dir}/deps/generic`). Cachi2 will verify that the target locations,
+including those derived from download urls do not overlap.
+
+### SBOM components
+
+Artifacts fetched with the generic fetcher will all be recorded in the SBOM cachi2 produces. Given the inability to derive
+any extra information about these files beyond a download location and a filename, these files will always be recorded
+as SBOM components with purl of type generic.
+
+Additionally, the SBOM component will contain [externalReferences] of type `distribution` to indicate the url used to download
+the file to allow for easier handling for tools that might process the SBOM.
+
+Here's an example SBOM generated for above file.
+
+```json
+{
+  "bomFormat": "CycloneDX",
+  "components": [
+    {
+      "name": "granite-model-1.safetensors",
+      "purl": "pkg:generic/granite-model-1.safetensors?checksums=sha256:d16bf783cb6670f7f692ad7d6885ab957c63cfc1b9649bc4a3ba1cfbdfd5230c&download_url=https://huggingface.co/instructlab/granite-7b-lab/resolve/main/model-00001-of-00003.safetensors",
```

Comment:
>  the strongest algorithm among the ones that were valid. This might be a good resolution for the matter here.

Do we want to get into the business of comparing and ordering hash algs based on how strong we think they are? Shouldn't this be in the user's own interest to provide strong algorithms which it wouldn't matter which one would be picked?

That said, remind me, do we really want N checksums per-artifact and validate all of them? It's going to slow down the whole process and I wonder whether the benefits outweigh the slight performance penalty paid.

@brunoapimentel  What is the inherent problem with reporting N checksums in a purl? The whole SBOM is a write only machine readable artifact anyway, so why does it matter how many checksums appear in there?

Reply:
> Do we want to get into the business of comparing and ordering hash algs based on how strong we think they are

Fair point, we could only pick the first valid one in the lockfile.

> That said, remind me, do we really want N checksums per-artifact and validate all of them? It's going to slow down the whole process and I wonder whether the benefits outweigh the slight performance penalty paid.

I believe the use cases we already have identified based on Maven have several checksums per artifact. We can try to confirm this. In any case, if we believe a single checksum is enough, I'd say the lockfile should support only a single one.

>  The whole SBOM is a write only machine readable artifact anyway, so why does it matter how many checksums appear in there?

I was looking at cases where the SBOM was being read by humans, but I agree with you it is the minority of cases. I'd be ok with having many checksums in the purls, but I still think only the valid ones should be there.

Reply:
> > Do we want to get into the business of comparing and ordering hash algs based on how strong we think they are

> Fair point, we could only pick the first valid one in the lockfile.

This is a potentially bad approach. I know folks in EC plan on gating content based on a list of allowed algos, so the reporting one at random, as you suggest, might potentially end up creating an unnecessary problem.  

> I believe the use cases we already have identified based on Maven have several checksums per artifact. We can try to confirm this. In any case, if we believe a single checksum is enough, I'd say the lockfile should support only a single one.

I agree with this. This will let user select what they think is the most appropriate algo, so any potential gating failure will be directly caused by them, not cachi2's algo selection heurestic.  I will go with this approach, and implement it so it's a string of "alg:hash", like this:

```
artifacts:
  - download_url: https://huggingface.co/instructlab/granite-7b-lab/resolve/main/model-00001-of-00003.safetensors?download=true
    filename: granite-model-1.safetensors
    checksum: "sha256:d16bf783cb6670f7f692ad7d6885ab957c63cfc1b9649bc4a3ba1cfbdfd5230c"
```

Reply:
> if we believe a single checksum is enough, I'd say the lockfile should support only a single one.

I understand that specifying a list of checksums and letting a different tool further down a pipeline to choose from but if we filter out SHA1 then having N hashes doesn't really bring any more confidence than a single hash out of the family of recommended strong algorithms, so why bother supporting that, it doesn't improve either the quality of the SBOM or provide any more confidence in it than going with a single one.

Reply:
Updated to only support a single checksum.